File: android/aspectly-bridge/src/main/kotlin/com/aspectly/bridge/BridgeHost.kt
package com.aspectly.bridge

import com.aspectly.bridge.protocol.BridgeErrorType
import com.aspectly.bridge.protocol.BridgeEventType
import com.aspectly.bridge.protocol.BridgeResultType
import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonNull
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSyntaxException
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout
import java.lang.reflect.Type
import java.util.concurrent.atomic.AtomicInteger

/**
 * `BridgeHost` manages bidirectional communication between native Android
 * (Kotlin) code and JavaScript via the Aspectly protocol. It handles message
 * passing in both directions and provides type-safe handler registration.
 *
 * This is the native (host) side of the bridge — the Kotlin equivalent of the
 * .NET `BridgeHost` and the parent side of the web hooks. The embedded web
 * content uses `@aspectly/core`'s `AspectlyBridge`.
 */
class BridgeHost @JvmOverloads constructor(
    private val browserBridge: BrowserBridge,
    private val logger: BridgeLogger = NullLogger,
    private val timeoutMs: Long = DEFAULT_TIMEOUT_MS,
) {
    companion object {
        /** Default request/handler timeout (100 seconds), matching `@aspectly/core`. */
        const val DEFAULT_TIMEOUT_MS = 100_000L
        private const val BRIDGE_EVENT_TYPE = "BridgeEvent"
    }

    private val gson = BridgeJson.gson
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val lock = Any()

    private var initialized = false
    private var remoteInitReceived = false
    private var initResultReceived = false
    private var disposed = false
    private val requestIdCounter = AtomicInteger(0)
    private var initDeferred: CompletableDeferred<Unit>? = null

    private val supportedMethodsStorage = mutableListOf<String>()
    private val handlers = mutableMapOf<String, RawBridgeHandler>()
    private val pendingRequests = mutableMapOf<String, CompletableDeferred<JsonElement>>()

    /** Whether the handshake (both `Init` and `InitResult`) has completed. */
    val isInitialized: Boolean
        get() = synchronized(lock) { initialized }

    /** The methods supported by the JavaScript side. */
    val supportedMethods: List<String>
        get() = synchronized(lock) { supportedMethodsStorage.toList() }

    /** The locally registered handler method names. */
    val registeredMethods: List<String>
        get() = synchronized(lock) { handlers.keys.toList() }

    /** Invoked when the bridge becomes initialized (handshake complete). */
    var onInitialized: (() -> Unit)? = null

    init {
        browserBridge.onMessage = { message -> scope.launch { processMessage(message) } }
        logger.info("[BridgeHost] Created and subscribed to messages")
    }

    // region Incoming messages

    /** Process a message received from JavaScript. */
    suspend fun processMessage(messageJson: String) {
        if (messageJson.isEmpty()) return

        logger.debug("[BridgeHost] Received: $messageJson")

        val root: JsonElement = try {
            JsonParser.parseString(messageJson)
        } catch (e: JsonSyntaxException) {
            logger.error("[BridgeHost] JSON parse error", e)
            return
        }

        val obj = root.asJsonObjectOrNull() ?: return
        if (obj.get("type")?.asStringOrNull() != BRIDGE_EVENT_TYPE) {
            logger.debug("[BridgeHost] Ignoring non-BridgeEvent message")
            return
        }

        val event = obj.get("event")?.asJsonObjectOrNull() ?: return
        val eventTypeRaw = event.get("type")?.asStringOrNull() ?: return
        val eventType = BridgeEventType.fromValue(eventTypeRaw) ?: return
        val data = event.get("data") ?: JsonNull.INSTANCE

        handleEvent(eventType, data)
    }

    private suspend fun handleEvent(type: BridgeEventType, data: JsonElement) {
        when (type) {
            BridgeEventType.INIT -> handleInit(data)
            BridgeEventType.REQUEST -> handleRequest(data)
            BridgeEventType.RESULT -> handleResult(data)
            BridgeEventType.INIT_RESULT -> {
                synchronized(lock) { initResultReceived = true }
                logger.info("[BridgeHost] InitResult received from JS")
                tryResolveInit()
            }
        }
    }

    private suspend fun handleInit(data: JsonElement) {
        val methods = data.asJsonObjectOrNull()
            ?.getAsJsonArrayOrNull("methods")
            ?.mapNotNull { it.asStringOrNull() }
        if (methods != null) {
            synchronized(lock) {
                supportedMethodsStorage.clear()
                supportedMethodsStorage.addAll(methods)
            }
            logger.info("[BridgeHost] JS supports methods: ${methods.joinToString(", ")}")
        }

        synchronized(lock) { remoteInitReceived = true }

        // Match JS protocol: only send InitResult, not our Init.
        // Our Init is sent explicitly via initialize().
        sendEvent(BridgeEventType.INIT_RESULT, JsonPrimitive(true))
        logger.info("[BridgeHost] Sent InitResult")
        tryResolveInit()
    }

    private fun tryResolveInit() {
        var deferred: CompletableDeferred<Unit>? = null
        var shouldResolve = false
        synchronized(lock) {
            if (initResultReceived && remoteInitReceived && !initialized) {
                initialized = true
                shouldResolve = true
                deferred = initDeferred
                initDeferred = null
            }
        }
        if (shouldResolve) {
            deferred?.complete(Unit)
            logger.info("[BridgeHost] Bridge fully initialized (both Init and InitResult received)")
            onInitialized?.invoke()
        }
    }

    private suspend fun handleRequest(data: JsonElement) {
        val obj = data.asJsonObjectOrNull() ?: return
        val method = obj.get("method")?.asStringOrNull() ?: return
        val requestId = obj.get("request_id")?.asStringOrNull() ?: return
        val params = obj.get("params") ?: JsonObject()

        val handler = synchronized(lock) { handlers[method] }

        val result: JsonObject = if (handler != null) {
            try {
                val handlerResult = withTimeout(timeoutMs) { handler(params) }
                logger.debug("[BridgeHost] Handler '$method' completed successfully")
                successResult(method, requestId, gson.toJsonTree(handlerResult))
            } catch (e: TimeoutCancellationException) {
                logger.error("[BridgeHost] Handler '$method' timed out after ${timeoutMs}ms")
                errorResult(method, requestId, BridgeErrorType.METHOD_EXECUTION_TIMEOUT, "Execution timeout exceeded")
            } catch (e: CancellationException) {
                throw e
            } catch (e: Throwable) {
                logger.error("[BridgeHost] Handler '$method' failed", e)
                errorResult(method, requestId, BridgeErrorType.REJECTED, e.message ?: e.toString())
            }
        } else {
            logger.warn("[BridgeHost] Unknown method: $method")
            errorResult(method, requestId, BridgeErrorType.UNSUPPORTED_METHOD, "Method '$method' is not registered")
        }

        sendEvent(BridgeEventType.RESULT, result)
    }

    private fun handleResult(data: JsonElement) {
        val obj = data.asJsonObjectOrNull() ?: return
        val requestId = obj.get("request_id")?.asStringOrNull() ?: return

        val deferred = synchronized(lock) { pendingRequests.remove(requestId) } ?: return

        if (obj.get("type")?.asStringOrNull() == BridgeResultType.SUCCESS.value) {
            deferred.complete(obj.get("data") ?: JsonNull.INSTANCE)
        } else {
            val error = obj.get("error")?.asJsonObjectOrNull()
            val errorType = error?.get("error_type")?.asStringOrNull()
                ?.let { BridgeErrorType.fromValue(it) } ?: BridgeErrorType.REJECTED
            val message = error?.get("error_message")?.asStringOrNull()
            deferred.completeExceptionally(BridgeException(errorType, message))
        }
    }

    // endregion

    // region Outgoing messages

    /** Send a raw event to JavaScript via `window.postMessage`. */
    suspend fun sendEvent(type: BridgeEventType, data: JsonElement) {
        val inner = JsonObject().apply {
            addProperty("type", type.value)
            add("data", data)
        }
        val wrapper = JsonObject().apply {
            addProperty("type", BRIDGE_EVENT_TYPE)
            add("event", inner)
        }

        val json = gson.toJson(wrapper)
        // Double-serialize to produce a safe JS string literal (matches .NET / Swift).
        val jsLiteral = gson.toJson(json)
        // Wrap in an IIFE returning a value so WebView.evaluateJavascript is happy.
        val script = "(function(){window.postMessage($jsLiteral, '*');return true;})();"

        try {
            browserBridge.executeScript(script)
            logger.debug("[BridgeHost] Sent: $json")
        } catch (e: Throwable) {
            logger.error("[BridgeHost] Failed to send event", e)
        }
    }

    /**
     * Send a request to the JavaScript side and await the typed response.
     * Prefer the `reified` overload for ergonomic call sites.
     */
    suspend fun <T> send(method: String, params: Any?, type: Type, timeoutMs: Long = DEFAULT_TIMEOUT_MS): T {
        if (!isInitialized) {
            throw BridgeException(BridgeErrorType.BRIDGE_NOT_AVAILABLE, "Bridge not initialized")
        }

        val supported = synchronized(lock) { supportedMethodsStorage.contains(method) }
        if (!supported) {
            throw BridgeException(BridgeErrorType.UNSUPPORTED_METHOD, "Method '$method' not supported by JS side")
        }

        val requestId = requestIdCounter.incrementAndGet().toString()
        val deferred = CompletableDeferred<JsonElement>()
        synchronized(lock) { pendingRequests[requestId] = deferred }

        val requestData = JsonObject().apply {
            addProperty("method", method)
            addProperty("request_id", requestId)
            add("params", gson.toJsonTree(params ?: JsonObject()))
        }

        sendEvent(BridgeEventType.REQUEST, requestData)
        logger.debug("[BridgeHost] Sent request: $method (id=$requestId)")

        val resultJson: JsonElement = try {
            withTimeout(timeoutMs) { deferred.await() }
        } catch (e: TimeoutCancellationException) {
            synchronized(lock) { pendingRequests.remove(requestId) }
            throw BridgeException(
                BridgeErrorType.METHOD_EXECUTION_TIMEOUT,
                "Request '$method' timed out after ${timeoutMs}ms"
            )
        }

        return gson.fromJson(resultJson, type)
    }

    /** Reified convenience for [send]. */
    suspend inline fun <reified T> send(
        method: String,
        params: Any? = null,
        timeoutMs: Long = DEFAULT_TIMEOUT_MS,
    ): T = send(method, params, object : TypeToken<T>() {}.type, timeoutMs)

    // endregion

    // region Handler registration

    /** Register a raw handler that receives the request params as a [JsonElement]. */
    fun registerHandler(method: String, handler: RawBridgeHandler) {
        synchronized(lock) { handlers[method] = handler }
        logger.info("[BridgeHost] Registered handler: $method")
    }

    /** Register a typed handler whose params are decoded into [P]. */
    inline fun <reified P> registerTypedHandler(
        method: String,
        crossinline handler: suspend (P) -> Any?,
    ) {
        registerHandler(method) { json -> handler(BridgeJson.gson.fromJson(json, P::class.java)) }
    }

    /** Register a handler implementing [BridgeHandler]. */
    fun <P, R> registerHandler(handler: BridgeHandler<P, R>, paramsType: Class<P>) {
        registerHandler(handler.methodName) { json -> handler.handle(BridgeJson.gson.fromJson(json, paramsType)) }
    }

    /** Remove a previously registered handler. */
    fun unregisterHandler(method: String) {
        synchronized(lock) { handlers.remove(method) }
        logger.info("[BridgeHost] Unregistered handler: $method")
    }

    // endregion

    // region Initialization

    /** Register handlers and send `Init` to JavaScript, then wait for `InitResult`. */
    suspend fun initialize(handlers: Map<String, RawBridgeHandler>) {
        handlers.forEach { (method, handler) -> registerHandler(method, handler) }
        initialize()
    }

    /**
     * Send the `Init` event to JavaScript with the registered methods and wait
     * for the handshake to complete. Mirrors the JS `bridge.init()` protocol.
     */
    suspend fun initialize() {
        val methods = registeredMethods
        val deferred = CompletableDeferred<Unit>()
        synchronized(lock) { initDeferred = deferred }

        val initData = JsonObject().apply {
            add("methods", JsonArray().apply { methods.forEach { add(it) } })
        }
        sendEvent(BridgeEventType.INIT, initData)
        logger.info("[BridgeHost] Sent Init with methods: ${methods.joinToString(", ")}")

        deferred.await()
        logger.info("[BridgeHost] InitResult received, initialization complete")
    }

    // endregion

    /** Detach from the browser bridge and cancel any pending work. */
    fun dispose() {
        val pending: List<CompletableDeferred<JsonElement>>
        val initCont: CompletableDeferred<Unit>?
        synchronized(lock) {
            if (disposed) return
            disposed = true
            pending = pendingRequests.values.toList()
            pendingRequests.clear()
            initCont = initDeferred
            initDeferred = null
        }

        browserBridge.onMessage = null
        browserBridge.dispose()

        val cancellation = CancellationException("BridgeHost disposed")
        initCont?.completeExceptionally(cancellation)
        pending.forEach { it.completeExceptionally(cancellation) }
        scope.cancel()

        logger.info("[BridgeHost] Disposed")
    }

    // region Helpers

    private fun successResult(method: String, requestId: String, data: JsonElement): JsonObject =
        JsonObject().apply {
            addProperty("type", BridgeResultType.SUCCESS.value)
            addProperty("method", method)
            addProperty("request_id", requestId)
            add("data", data)
        }

    private fun errorResult(
        method: String,
        requestId: String,
        errorType: BridgeErrorType,
        message: String,
    ): JsonObject = JsonObject().apply {
        addProperty("type", BridgeResultType.ERROR.value)
        addProperty("method", method)
        addProperty("request_id", requestId)
        add("error", JsonObject().apply {
            addProperty("error_type", errorType.value)
            addProperty("error_message", message)
        })
    }

    // endregion
}

private fun JsonElement.asJsonObjectOrNull(): JsonObject? =
    if (isJsonObject) asJsonObject else null

private fun JsonObject.getAsJsonArrayOrNull(key: String): JsonArray? =
    get(key)?.takeIf { it.isJsonArray }?.asJsonArray

private fun JsonElement.asStringOrNull(): String? =
    if (isJsonPrimitive && asJsonPrimitive.isString) asString else null

File: dart/lib/src/bridge_host.dart
import 'dart:async';
import 'dart:convert';

import 'browser_bridge.dart';
import 'bridge_exception.dart';
import 'protocol.dart';

/// A handler for an incoming request. Receives the decoded params and returns
/// a JSON-encodable result (sync or async).
typedef BridgeHandler = FutureOr<dynamic> Function(dynamic params);

/// `BridgeHost` manages bidirectional communication between native Dart code
/// and JavaScript via the Aspectly protocol. The Dart equivalent of the .NET
/// `BridgeHost` and the parent side of the web hooks.
class BridgeHost {
  /// Default request/handler timeout (100 seconds), matching `@aspectly/core`.
  static const int defaultTimeoutMs = 100000;
  static const String _bridgeEventType = 'BridgeEvent';

  final BrowserBridge _browserBridge;
  final BridgeLogger _logger;
  final int _timeoutMs;

  bool _initialized = false;
  bool _remoteInitReceived = false;
  bool _initResultReceived = false;
  bool _disposed = false;
  int _requestIdCounter = 0;
  Completer<void>? _initCompleter;

  final List<String> _supportedMethods = <String>[];
  final Map<String, BridgeHandler> _handlers = <String, BridgeHandler>{};
  final Map<String, Completer<dynamic>> _pendingRequests =
      <String, Completer<dynamic>>{};

  /// Invoked when the bridge becomes initialized (handshake complete).
  void Function()? onInitialized;

  BridgeHost(
    this._browserBridge, {
    BridgeLogger? logger,
    int timeoutMs = defaultTimeoutMs,
  })  : _logger = logger ?? const NullLogger(),
        _timeoutMs = timeoutMs {
    _browserBridge.onMessage = (String message) {
      // Fire-and-forget; processMessage handles its own errors.
      processMessage(message);
    };
    _logger.info('[BridgeHost] Created and subscribed to messages');
  }

  /// Whether the handshake (both `Init` and `InitResult`) has completed.
  bool get isInitialized => _initialized;

  /// The methods supported by the JavaScript side.
  List<String> get supportedMethods => List<String>.unmodifiable(_supportedMethods);

  /// The locally registered handler method names.
  List<String> get registeredMethods => List<String>.unmodifiable(_handlers.keys);

  /// Register a handler for [method]. Call before [initialize].
  void registerHandler(String method, BridgeHandler handler) {
    _handlers[method] = handler;
    _logger.info('[BridgeHost] Registered handler: $method');
  }

  /// Remove a previously registered handler.
  void unregisterHandler(String method) {
    _handlers.remove(method);
    _logger.info('[BridgeHost] Unregistered handler: $method');
  }

  /// Process a message received from JavaScript.
  Future<void> processMessage(String messageJson) async {
    if (messageJson.isEmpty) return;
    _logger.debug('[BridgeHost] Received: $messageJson');

    dynamic wrapper;
    try {
      wrapper = jsonDecode(messageJson);
    } catch (e) {
      _logger.error('[BridgeHost] JSON parse error', e);
      return;
    }

    if (wrapper is! Map || wrapper['type'] != _bridgeEventType) return;
    final dynamic event = wrapper['event'];
    if (event is! Map) return;
    final dynamic type = event['type'];
    final dynamic data = event['data'];

    switch (type) {
      case BridgeEventType.init:
        await _handleInit(data);
        break;
      case BridgeEventType.request:
        await _handleRequest(data);
        break;
      case BridgeEventType.result:
        _handleResult(data);
        break;
      case BridgeEventType.initResult:
        _initResultReceived = true;
        _logger.info('[BridgeHost] InitResult received from JS');
        _tryResolveInit();
        break;
    }
  }

  Future<void> _handleInit(dynamic data) async {
    if (data is Map && data['methods'] is List) {
      _supportedMethods
        ..clear()
        ..addAll((data['methods'] as List).map((dynamic e) => e.toString()));
      _logger.info('[BridgeHost] JS supports methods: ${_supportedMethods.join(', ')}');
    }
    _remoteInitReceived = true;
    // Match the JS protocol: only send InitResult, not our Init.
    await _sendEvent(BridgeEventType.initResult, true);
    _logger.info('[BridgeHost] Sent InitResult');
    _tryResolveInit();
  }

  void _tryResolveInit() {
    if (_initResultReceived && _remoteInitReceived && !_initialized) {
      _initialized = true;
      final Completer<void>? completer = _initCompleter;
      _initCompleter = null;
      completer?.complete();
      _logger.info('[BridgeHost] Bridge fully initialized');
      final cb = onInitialized;
      if (cb != null) cb();
    }
  }

  Future<void> _handleRequest(dynamic data) async {
    if (data is! Map) return;
    final String? method = data['method'] as String?;
    final String? requestId = data['request_id'] as String?;
    if (method == null || requestId == null) return;
    final dynamic params = data.containsKey('params') ? data['params'] : <String, dynamic>{};

    final BridgeHandler? handler = _handlers[method];
    Map<String, dynamic> result;
    if (handler == null) {
      _logger.warn('[BridgeHost] Unknown method: $method');
      result = _errorResult(method, requestId, BridgeErrorType.unsupportedMethod,
          "Method '$method' is not registered");
    } else {
      try {
        final dynamic value = await Future<dynamic>(() => handler(params))
            .timeout(Duration(milliseconds: _timeoutMs));
        result = <String, dynamic>{
          'type': BridgeResultType.success,
          'method': method,
          'request_id': requestId,
          'data': value,
        };
        _logger.debug("[BridgeHost] Handler '$method' completed successfully");
      } on TimeoutException {
        _logger.error("[BridgeHost] Handler '$method' timed out");
        result = _errorResult(method, requestId,
            BridgeErrorType.methodExecutionTimeout, 'Execution timeout exceeded');
      } catch (e) {
        _logger.error("[BridgeHost] Handler '$method' failed", e);
        final String msg = e is BridgeException ? e.message : e.toString();
        result = _errorResult(method, requestId, BridgeErrorType.rejected, msg);
      }
    }
    await _sendEvent(BridgeEventType.result, result);
  }

  void _handleResult(dynamic data) {
    if (data is! Map) return;
    final String? requestId = data['request_id'] as String?;
    if (requestId == null) return;
    final Completer<dynamic>? completer = _pendingRequests.remove(requestId);
    if (completer == null || completer.isCompleted) return;

    if (data['type'] == BridgeResultType.success) {
      completer.complete(data['data']);
    } else {
      final dynamic error = data['error'];
      final String errorType =
          (error is Map ? error['error_type'] : null)?.toString() ?? BridgeErrorType.rejected;
      final String? message = (error is Map ? error['error_message'] : null)?.toString();
      completer.completeError(BridgeException(errorType, message));
    }
  }

  Future<void> _sendEvent(String type, dynamic data) async {
    final Map<String, dynamic> wrapper = <String, dynamic>{
      'type': _bridgeEventType,
      'event': <String, dynamic>{'type': type, 'data': data},
    };
    final String json = jsonEncode(wrapper);
    // Double-encode to produce a safe JS string literal (matches .NET / Swift).
    final String jsLiteral = jsonEncode(json);
    final String script =
        "(function(){window.postMessage($jsLiteral, '*');return true;})();";
    try {
      await _browserBridge.executeScript(script);
      _logger.debug('[BridgeHost] Sent: $json');
    } catch (e) {
      _logger.error('[BridgeHost] Failed to send event', e);
    }
  }

  /// Send a request to the JavaScript side and await the decoded response.
  Future<T> send<T>(String method, [dynamic params, int? timeoutMs]) async {
    if (!_initialized) {
      throw BridgeException(BridgeErrorType.bridgeNotAvailable, 'Bridge not initialized');
    }
    if (!_supportedMethods.contains(method)) {
      throw BridgeException(BridgeErrorType.unsupportedMethod,
          "Method '$method' not supported by JS side");
    }

    final String requestId = (++_requestIdCounter).toString();
    final Completer<dynamic> completer = Completer<dynamic>();
    _pendingRequests[requestId] = completer;

    await _sendEvent(BridgeEventType.request, <String, dynamic>{
      'method': method,
      'request_id': requestId,
      'params': params ?? <String, dynamic>{},
    });
    _logger.debug('[BridgeHost] Sent request: $method (id=$requestId)');

    final Timer timer = Timer(Duration(milliseconds: timeoutMs ?? _timeoutMs), () {
      final Completer<dynamic>? c = _pendingRequests.remove(requestId);
      if (c != null && !c.isCompleted) {
        c.completeError(BridgeException(BridgeErrorType.methodExecutionTimeout,
            "Request '$method' timed out after ${timeoutMs ?? _timeoutMs}ms"));
      }
    });

    try {
      final dynamic result = await completer.future;
      return result as T;
    } finally {
      timer.cancel();
    }
  }

  /// Register handlers and send `Init`, then wait for `InitResult`.
  Future<void> initialize([Map<String, BridgeHandler>? handlers]) async {
    if (handlers != null) {
      handlers.forEach(registerHandler);
    }
    final Completer<void> completer = Completer<void>();
    _initCompleter = completer;
    await _sendEvent(BridgeEventType.init, <String, dynamic>{
      'methods': _handlers.keys.toList(),
    });
    _logger.info('[BridgeHost] Sent Init with methods: ${_handlers.keys.join(', ')}');
    await completer.future;
  }

  /// Detach from the browser bridge and cancel any pending work.
  void dispose() {
    if (_disposed) return;
    _disposed = true;
    _browserBridge.onMessage = null;
    _browserBridge.dispose();

    final BridgeException cancellation =
        BridgeException(BridgeErrorType.bridgeNotAvailable, 'Bridge disposed');
    final Completer<void>? initCompleter = _initCompleter;
    _initCompleter = null;
    if (initCompleter != null && !initCompleter.isCompleted) {
      initCompleter.completeError(cancellation);
    }
    for (final Completer<dynamic> c in _pendingRequests.values) {
      if (!c.isCompleted) c.completeError(cancellation);
    }
    _pendingRequests.clear();
    _logger.info('[BridgeHost] Disposed');
  }

  Map<String, dynamic> _errorResult(
      String method, String requestId, String errorType, String message) {
    return <String, dynamic>{
      'type': BridgeResultType.error,
      'method': method,
      'request_id': requestId,
      'error': <String, dynamic>{
        'error_type': errorType,
        'error_message': message,
      },
    };
  }
}
